File: solutions/cpp/1631/solution.cpp
class Solution {
    const vector<vector<int>> DIRECTIONS{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    int ROWS;
    int COLS;
    vector<vector<int>> heights;
    int minimumEffortPath(vector<int> &heights) {
        this->heights = heights;
        ROWS = heights.size();
        COLS = heights[0].size();

        // First let's get our search space
        int left = 0, right = 0;
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < ROWS; j++) {
                right = max(right, heights[i][j]);
            }
        }

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (check(mid)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    bool check(int effort) {
        vector<vector<bool>> seen(ROWS, vector<bol>(COLS, false));
        seen[0][0] = true;
        dfs(0, 0, effort, seen);
        return seen[ROWS - 1][COLS - 1];
    }

    void dfs(int row, int col, int effort, vector<vector<bool>> &seen) {
        for (auto &dir : DIRECTIONS) {
            int newRow = row + dir[0];
            int newCol = col + dir[1];
            if (valid(newRow, newCol) && !seen[newRow][newCol]) {
                if (abs(heights[newRow][newCol] - heights[row][col]) <= effort) {
                    seen[newRow][newCol] = true;
                    dfs(newRow, newCol, effort, seen);
                }
            }
        }
    }

    bool valid(int row, int col) {
        return row >= 0 && row < ROWS && col >= 0 && col < COLS;
    }
}

File: solutions/golang/1631/solution.go
var directions = [][]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}
var M int
var N int
func minimumEffortPath(heights [][]int) int {
    left := 0
    right := 0
    M = len(heights)
    N = len(heights[0])
    for row := 0; row < M; row++ {
        for col := 0; col < N; col++ {
            right = max(right, heights[row][col])
        }
    }

    for left <= right {
        mid := left + (right - left) / 2
        if check(mid, heights) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return left
}

func check(effort int, heights [][]int) bool {
    seen := make([][]bool, M)
    for i := 0; i < M; i++ {
        seen[i] = make([]bool, N)
    }

    seen[0][0] = true
    dfs(0, 0, effort, heights, seen)

    return seen[M-1][N-1]
}

func dfs(row int, col int, effort int, heights [][]int, seen [][]bool) {
    for _, dir := range directions {
        newRow := row + dir[0]
        newCol := col + dir[1]
        if valid(newRow, newCol) && !seen[newRow][newCol] {
            newEffort := heights[newRow][newCol] - heights[row][col]
            if newEffort < 0 {
                newEffort = -newEffort
            }
            if newEffort <= effort {
                seen[newRow][newCol] = true
                dfs(newRow, newCol, effort, heights, seen)
            }
        }
    }
}

func valid(row int, col int) bool {
    return row >= 0 && row < M && col >= 0 && col < N
}
